### PR TITLE
value: Add a Sliceable interface for operands of the [] operator

### LINKED ISSUE #104: feature request: allow custom skylark Values to implement the slice operator
in real Python, user-defined classes can implement the slice operator (`x[i:j]`)

in Skylark, the slice operator can only be used on String, List, and Tuple
https://github.com/google/skylark/blob/master/eval.go#L883

would you accept a PR to make this an interface, like other Skylark operators (HasAttrs, etc)?

### PR DESCRIPTION
**Pull request by nicks**
Fixes #104

### CODE REVIEW

**Comment by nicks** [comment]
Friendly ping, in case this got missed over the weekend

File: value.go [hunk 1]
```diff
@@ -179,6 +179,18 @@ type Indexable interface {
...
 
+// A Sliceable is a sequence that can be cut into pieces with the slice operator (x[i:j:step]).
+//
+// All native indexable objects are sliceable.
+// This is a separate interface for backwards-compatibility.
+type Sliceable interface {
+	Indexable
+	// For positive strides (step > 0), start < end.
+	// For negative strides (step < 0), start > end.
+	// The implementation of Slice should treat start as inclusive end as non-inclusive.
```
Comment (alandonovan) [inline value.go]: This assumption permeates all of Skylark (and all of Go). Strike this sentence.

File: value.go [hunk 2]
```diff
@@ -179,6 +179,18 @@ type Indexable interface {
 	Len() int
 }
 
+// A Sliceable is a sequence that can be cut into pieces with the slice operator (x[i:j:step]).
+//
+// All native indexable objects are sliceable.
+// This is a separate interface for backwards-compatibility.
+type Sliceable interface {
+	Indexable
+	// For positive strides (step > 0), start < end.
```
Comment (alandonovan) [inline value.go]: The inequalities here should be start <= end, and start >= end on the next line.

More precisely:
    0 <= start <= end <= n
    -1 <= end <= start < n

File: eval.go [hunk 3]
```diff
@@ -876,31 +863,11 @@ func slice(x, lo, hi, step_ Value) (Value, error) {
...
-		}
-		return NewList(list), nil
-	case Tuple:
-		var tuple Tuple
-		for i := start; signum(end-i) == sign; i += step {
-			tuple = append(tuple, x[i])
-		}
-		return tuple, nil
+	sliceable, ok := x.(Sliceable)
+	if !ok {
```
Comment (alandonovan) [inline eval.go]: Let's simplify by moving this check up to the start:

	func slice(x, lo, hi, step_ Value) (Value, error) {
		sliceable, ok := x.(Sliceable)
		if !ok {
	 		return nil, fmt.Errorf("invalid slice operand %s", x.Type())
		}
		n := sliceable.Len()

We may assume n is nonnegative.

**Review by alandonovan — COMMENTED** [review]
Thanks, nice change.

While you're here, would you mind making the data type returned by range(n) implement Sliceable, following Python3 semantics?

For example:
```
>>> range(10)[1:]
range(1, 10)
>>> range(10)[1:9:2]
range(1, 9, 2)
```

**Comment by nicks** [comment]
all comments addressed. Made rangeValue a Sliceable and added a bunch of tests.

**Review by alandonovan — APPROVED** [review]
Perfect, thanks.

**Pull request merged**